### PR TITLE
Crosslink rewards

### PR DESCRIPTION
**Pull request by djrtwo**
addresses #119 

- Fills out `calculate_crosslink_rewards` function
- need more testing

### CODE REVIEW

File: beacon_chain/state/state_transition.py [hunk 1]
```diff
@@ -484,10 +488,77 @@ def calculate_crosslink_rewards(crystallized_state: CrystallizedState,
...
-    #
-    # STUB
-    # Still need clarity in spec to properly fill these calculations
-    #
+    total_deposits = crystallized_state.total_deposits
+    total_deposits_in_ETH = total_deposits // WEI_PER_ETH
+    reward_quotient = config['base_reward_quotient'] * int(sqrt(total_deposits_in_ETH))
+    quadratic_penalty_quotient = int(sqrt(config['sqrt_e_drop_time'] / config['slot_duration']))
+
+    last_state_recalc = crystallized_state.last_state_recalc
```
Comment (hwwhww) [inline beacon_chain/state/state_transition.py]: This section is a duplicate of `calculate_ffg_rewards`.

File: beacon_chain/state/state_transition.py [hunk 2]
```diff
@@ -484,10 +488,77 @@ def calculate_crosslink_rewards(crystallized_state: CrystallizedState,
...
+            if attestations:
+                bitfields = [
+                    attestation.attester_bitfield for attestation in attestations
+                ]
+                bitfield = or_bitfields(bitfields)
+            else:
+                bitfield = get_empty_bitfield(len(shard_and_committee.committee))
+
+            committee_crosslink = committee_crosslinks[shard_id]
+            for i, index in enumerate(shard_and_committee.committee):
```
Comment (hwwhww) [inline beacon_chain/state/state_transition.py]: IMO `i` and `index` are confusing. What do you think of renaming `index` to `validator_index`?

Comment (djrtwo) [reply beacon_chain/state/state_transition.py]: going to go with `i` and `validator_index`.

Comment (djrtwo) [reply beacon_chain/state/state_transition.py]: actually, I went with `committee_index, validator_index`

**Comment by djrtwo** [comment]
wrote some tests that uncovered potential bug in the spec. Need to leave this as WIP until I get some clarity from V

File: beacon_chain/state/state_transition.py [hunk 3]
```diff
@@ -511,10 +515,79 @@ def calculate_crosslink_rewards(crystallized_state: CrystallizedState,
...
+                else:
+                    committee_crosslink[
+                        'non_participating_validator_indices'
+                    ].append(validator_index)
+
+                committee_crosslink['total_v_deposits'] += validator.balance
+
+    # for each shard and associated validator set, apply rewards/penalties based on participation
+    for shard_id in committee_crosslinks:
+        committee_crosslink = committee_crosslinks[shard_id]
```
[the remark this answers is not in the document]
Comment (djrtwo) [reply beacon_chain/state/state_transition.py]: Didn't know `items()` existed. Yes!

File: tests/state/test_state_transition.py [hunk 4]
```diff
@@ -266,6 +269,69 @@ def test_validate_attestation_aggregate_sig(attestation_validation_fixture, conf
...
+                                     mock_make_attestations,
+                                     mock_make_child):
+    c = genesis_crystallized_state
+    a = genesis_active_state
+    block = genesis_block
+    a.chain = Chain(head=block, blocks=[block])
+
+    # progress past first cycle transition
+    # rewards on the following cycle recalc will be based
+    # on whhat happened during this cycle
```
[the remark this answers is not in the document]
Comment (djrtwo) [reply tests/state/test_state_transition.py]: whhat?

**Pull request merged**